From 2e511d515011062a2ec0540ca14ebdef4c70abb3 Mon Sep 17 00:00:00 2001
From: ricolin <rlin@vexxhost.com>
Date: Mon, 24 Mar 2025 15:34:18 +0800
Subject: [PATCH 1/9] Add Octavia OVN DBs sync cmd

This is initial patch of a chain to allow ovn_octavia_provider
to be able to sync Octavia DB to OVN NB DB to sync content,
getting as source of true the Octavia DB one.

This patch add command sync command `octavia-ovn-db-sync-util`

Related-Bug: #2045415

Co-authored-by: Fernando Royo <froyo@redhat.com>
Co-authored-by: Rico Lin <ricolin@ricolky.com>

Change-Id: I0b1fee7a75e0a2a837e89766ea3c3198e0929823
---
 etc/octavia/conf.d/ovn.conf.sample            | 86 +++++++++++++++++++
 ovn_octavia_provider/cmd/__init__.py          |  0
 .../cmd/octavia_ovn_db_sync_util.py           | 59 +++++++++++++
 ovn_octavia_provider/driver.py                |  5 ++
 setup.cfg                                     |  3 +
 5 files changed, 153 insertions(+)
 create mode 100644 etc/octavia/conf.d/ovn.conf.sample
 create mode 100644 ovn_octavia_provider/cmd/__init__.py
 create mode 100644 ovn_octavia_provider/cmd/octavia_ovn_db_sync_util.py

diff --git a/etc/octavia/conf.d/ovn.conf.sample b/etc/octavia/conf.d/ovn.conf.sample
new file mode 100644
index 0000000..41fba46
--- /dev/null
+++ b/etc/octavia/conf.d/ovn.conf.sample
@@ -0,0 +1,86 @@
+[DEFAULT]
+
+
+[neutron]
+
+#
+# From octavia.api.drivers.ovn
+#
+
+# DEPRECATED: A new endpoint to override the endpoint in the keystone catalog.
+# (string value)
+# This option is deprecated for removal since Antelope.
+# Its value may be silently ignored in the future.
+# Reason: The endpoint_override option defined by keystoneauth1 is the new name
+# for this option.
+#endpoint = <None>
+
+# DEPRECATED: Endpoint interface in identity service to use (string value)
+# This option is deprecated for removal since Antelope.
+# Its value may be silently ignored in the future.
+# Reason: This option was replaced by the valid_interfaces option defined by
+# keystoneauth.
+#endpoint_type = <None>
+
+# DEPRECATED: CA certificates file path (string value)
+# This option is deprecated for removal since Antelope.
+# Its value may be silently ignored in the future.
+# Reason: The cafile option defined by keystoneauth1 is the new name for this
+# option.
+#ca_certificates_file = <None>
+
+
+[ovn]
+
+#
+# From octavia.api.drivers.ovn
+#
+
+# The connection string for the OVN_Northbound OVSDB.
+# Use tcp:IP:PORT for TCP connection.
+# Use ssl:IP:PORT for SSL connection. The ovn_nb_private_key,
+# ovn_nb_certificate and ovn_nb_ca_cert are mandatory.
+# Use unix:FILE for unix domain socket connection. (string value)
+#ovn_nb_connection = tcp:127.0.0.1:6641
+
+# The PEM file with private key for SSL connection to OVN-NB-DB (string value)
+#ovn_nb_private_key =
+
+# The PEM file with certificate that certifies the private key specified in
+# ovn_nb_private_key (string value)
+#ovn_nb_certificate =
+
+# The PEM file with CA certificate that OVN should use to verify certificates
+# presented to it by SSL peers (string value)
+#ovn_nb_ca_cert =
+
+# The connection string for the OVN_Southbound OVSDB.
+# Use tcp:IP:PORT for TCP connection.
+# Use ssl:IP:PORT for SSL connection. The ovn_sb_private_key,
+# ovn_sb_certificate and ovn_sb_ca_cert are mandatory.
+# Use unix:FILE for unix domain socket connection. (string value)
+#ovn_sb_connection = tcp:127.0.0.1:6642
+
+# The PEM file with private key for SSL connection to OVN-SB-DB (string value)
+#ovn_sb_private_key =
+
+# The PEM file with certificate that certifies the private key specified in
+# ovn_sb_private_key (string value)
+#ovn_sb_certificate =
+
+# The PEM file with CA certificate that OVN should use to verify certificates
+# presented to it by SSL peers (string value)
+#ovn_sb_ca_cert =
+
+# Timeout in seconds for the OVSDB connection transaction (integer value)
+#ovsdb_connection_timeout = 180
+
+# Max interval in seconds between each retry to get the OVN NB and SB IDLs
+# (integer value)
+#ovsdb_retry_max_interval = 180
+
+# The probe interval in for the OVSDB session in milliseconds. If this is zero,
+# it disables the connection keepalive feature. If non-zero the value will be
+# forced to at least 1000 milliseconds. Defaults to 60 seconds. (integer value)
+# Minimum value: 0
+#ovsdb_probe_interval = 60000
diff --git a/ovn_octavia_provider/cmd/__init__.py b/ovn_octavia_provider/cmd/__init__.py
new file mode 100644
index 0000000..e69de29
diff --git a/ovn_octavia_provider/cmd/octavia_ovn_db_sync_util.py b/ovn_octavia_provider/cmd/octavia_ovn_db_sync_util.py
new file mode 100644
index 0000000..b42e7ac
--- /dev/null
+++ b/ovn_octavia_provider/cmd/octavia_ovn_db_sync_util.py
@@ -0,0 +1,59 @@
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+import sys
+
+from oslo_config import cfg
+from oslo_log import log as logging
+from ovn_octavia_provider.common import config as ovn_conf
+from ovn_octavia_provider import driver
+
+CONF = cfg.CONF
+LOG = logging.getLogger(__name__)
+
+
+def setup_conf():
+    conf = cfg.CONF
+    ovn_conf.register_opts()
+    logging.register_options(CONF)
+
+    try:
+        CONF(project='octavia')
+    except TypeError:
+        LOG.error('Error parsing the configuration values. Please verify.')
+        raise
+    return conf
+
+
+def main():
+    """Main method for syncing Octavia LBs (OVN provider) with OVN NB DB.
+
+    This script provides a utility for syncing the OVN Northbound Database
+    with the Octavia database.
+
+    """
+    setup_conf()
+    logging.setup(CONF, 'octavia_ovn_db_sync_util')
+
+    # Method can be call like `octavia-ovn-db-sync-util --debug`
+    LOG.info("OVN Octavia DB sync start.")
+    args = sys.argv[1:]
+    lb_filters = {'provider': 'ovn'}
+    if '--debug' in args:
+        cfg.CONF.set_override('debug', True)
+        args.remove('--debug')
+    else:
+        cfg.CONF.set_override('debug', False)
+
+    ovn_driver = driver.OvnProviderDriver()
+    ovn_driver.do_sync(**lb_filters)
+    LOG.info("OVN Octavia DB sync finish.")
diff --git a/ovn_octavia_provider/driver.py b/ovn_octavia_provider/driver.py
index 947e280..ae225b6 100644
--- a/ovn_octavia_provider/driver.py
+++ b/ovn_octavia_provider/driver.py
@@ -533,3 +533,8 @@ class OvnProviderDriver(driver_base.ProviderDriver):
         request = {'type': ovn_const.REQ_TYPE_HM_DELETE,
                    'info': request_info}
         self._ovn_helper.add_request(request)
+
+    def do_sync(self, **lb_filters):
+        LOG.info(f"Starting sync OVN DB with Loadbalancer filter {lb_filters}")
+        # TODO(froyo): get LBs from Octavia DB through openstack sdk client and
+        # call to helper methods to sync
diff --git a/setup.cfg b/setup.cfg
index 06a3c0d..c883b32 100644
--- a/setup.cfg
+++ b/setup.cfg
@@ -30,6 +30,9 @@ setup_hooks =
 octavia.api.drivers =
     ovn = ovn_octavia_provider.driver:OvnProviderDriver
 
+console_scripts =
+    octavia-ovn-db-sync-util = ovn_octavia_provider.cmd.octavia_ovn_db_sync_util:main
+
 octavia.driver_agent.provider_agents =
     ovn = ovn_octavia_provider.agent:OvnProviderAgent
 
-- 
2.25.1

